File: 0543_diameter_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.maximum = 0
        
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return
        left = root.left
        right = root.right
        self.maximum = max(self.maximum, self.calculate(left) + self.calculate(right))
        self.diameterOfBinaryTree(root.left)
        self.diameterOfBinaryTree(root.right)
        return self.maximum
        
    def calculate(self, node):
        if not node:
            return 0
        return 1 + max(self.calculate(node.left), self.calculate(node.right))
      
# Alternative solution

class Solution1:
    def __init__(self):
        self.maximum = 0
        self.hashtable = {}
        
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.calculate(root)
        self.dfs(root)
        return self.maximum        
    
    def dfs(self, node):
        if not node.left and not node.right:
            return 0
        elif not node.right:
            self.maximum = max(self.maximum, self.hashtable[node.left])
        elif not node.left:
            self.maximum = max(self.maximum, self.hashtable[node.right])
        else:
            self.maximum = max(self.maximum, self.hashtable[node.left] + self.hashtable[node.right])
        if node.left:
            self.dfs(node.left)
        if node.right:
            self.dfs(node.right)
        return
        
    def calculate(self, node):
        if not node:
            return 0
        val = 1 + max(self.calculate(node.left), self.calculate(node.right))
        self.hashtable[node] = val
        return val

File: 0501-1000/0543_diameterOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return diameter;
    }

    public int dfs(TreeNode root) {
        if (root.left == null && root.right == null) {
            return 0;
        }

        int left = (root.left != null) ? dfs(root.left) + 1 : 0;
        int right = (root.right != null) ? dfs(root.right) + 1 : 0;
        diameter = Math.max(diameter, left + right);
        return Math.max(left, right);
    }
}
